In [76]:

# IMPORTANT: RUN THIS CELL IN ORDER TO IMPORT YOUR KAGGLE DATA SOURCES
# TO THE CORRECT LOCATION (/kaggle/input) IN YOUR NOTEBOOK,
# THEN FEEL FREE TO DELETE THIS CELL.
# NOTE: THIS NOTEBOOK ENVIRONMENT DIFFERS FROM KAGGLE'S PYTHON
# ENVIRONMENT SO THERE MAY BE MISSING LIBRARIES USED BY YOUR
# NOTEBOOK.

import os
import sys
from tempfile import NamedTemporaryFile
from urllib.request import urlopen
from urllib.parse import unquote, urlparse
from urllib.error import HTTPError
from zipfile import ZipFile
import tarfile
import shutil

CHUNK_SIZE = 40960
DATA_SOURCE_MAPPING = 'bi-ml-competition-2023:https%3A%2F%2Fstorage.googleapis.com%2Fkaggle-competitions-data%2Fkaggle-v2%2F49431%2F5228520%2Fbundle%2Farchive.zip%3FX-Goog-Algorithm%3DGOOG4-RSA-SHA256%26X-Goog-Credential%3Dgcp-kaggle-com%2540kaggle-161607.iam.gserviceaccount.com%252F20240311%252Fauto%252Fstorage%252Fgoog4_request%26X-Goog-Date%3D20240311T140336Z%26X-Goog-Expires%3D259200%26X-Goog-SignedHeaders%3Dhost%26X-Goog-Signature%3D0439a6647118ef93fac3b8d2f72135cfd433fa2ff3d0992a9bb24360bf7e359ce6346e0acb775939df8562e3d5ce4efb67bf65176232bd7293a8398aa85ebc4751da621a3cacef50ff5146703ffdff98d954899ee2baa8cf917f55500c0917d2f532b56b4417aa1ec82004552736d80fdc551ea663815efdb030039529fb4bbe7f4cfc7a9d859ef8ec998a3d5e8792b96eedf04e359fa2cc87bc5a691514009a2b428db00d65db322390551a6651afabef44689b1d40bfd1670fa5ff224f377ce4821e5f3b4f1a39b14007b4a62c1855d6e2f7348d17370c1dca2a4a17aa19cfeb7286b7427a95c9ca95d1190483075c33f893db9fc8526f59e53fee57f34673'

KAGGLE_INPUT_PATH='/kaggle/input'
KAGGLE_WORKING_PATH='/kaggle/working'
KAGGLE_SYMLINK='kaggle'

!umount /kaggle/input/ 2> /dev/null
shutil.rmtree('/kaggle/input', ignore_errors=True)
os.makedirs(KAGGLE_INPUT_PATH, 0o777, exist_ok=True)
os.makedirs(KAGGLE_WORKING_PATH, 0o777, exist_ok=True)

try:
  os.symlink(KAGGLE_INPUT_PATH, os.path.join("..", 'input'), target_is_directory=True)
except FileExistsError:
  pass
try:
  os.symlink(KAGGLE_WORKING_PATH, os.path.join("..", 'working'), target_is_directory=True)
except FileExistsError:
  pass

for data_source_mapping in DATA_SOURCE_MAPPING.split(','):
    directory, download_url_encoded = data_source_mapping.split(':')
    download_url = unquote(download_url_encoded)
    filename = urlparse(download_url).path
    destination_path = os.path.join(KAGGLE_INPUT_PATH, directory)
    try:
        with urlopen(download_url) as fileres, NamedTemporaryFile() as tfile:
            total_length = fileres.headers['content-length']
            print(f'Downloading {directory}, {total_length} bytes compressed')
            dl = 0
            data = fileres.read(CHUNK_SIZE)
            while len(data) > 0:
                dl += len(data)
                tfile.write(data)
                done = int(50 * dl / int(total_length))
                sys.stdout.write(f"\r[{'=' * done}{' ' * (50-done)}] {dl} bytes downloaded")
                sys.stdout.flush()
                data = fileres.read(CHUNK_SIZE)
            if filename.endswith('.zip'):
              with ZipFile(tfile) as zfile:
                zfile.extractall(destination_path)
            else:
              with tarfile.open(tfile.name) as tarfile:
                tarfile.extractall(destination_path)
            print(f'\nDownloaded and uncompressed: {directory}')
    except HTTPError as e:
        print(f'Failed to load (likely expired) {download_url} to path {destination_path}')
        continue
    except OSError as e:
        print(f'Failed to load {download_url} to path {destination_path}')
        continue

print('Data source import complete.')


Failed to load https://storage.googleapis.com/kaggle-competitions-data/kaggle-v2/49431/5228520/bundle/archive.zip?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=gcp-kaggle-com%40kaggle-161607.iam.gserviceaccount.com%2F20240311%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20240311T140336Z&X-Goog-Expires=259200&X-Goog-SignedHeaders=host&X-Goog-Signature=0439a6647118ef93fac3b8d2f72135cfd433fa2ff3d0992a9bb24360bf7e359ce6346e0acb775939df8562e3d5ce4efb67bf65176232bd7293a8398aa85ebc4751da621a3cacef50ff5146703ffdff98d954899ee2baa8cf917f55500c0917d2f532b56b4417aa1ec82004552736d80fdc551ea663815efdb030039529fb4bbe7f4cfc7a9d859ef8ec998a3d5e8792b96eedf04e359fa2cc87bc5a691514009a2b428db00d65db322390551a6651afabef44689b1d40bfd1670fa5ff224f377ce4821e5f3b4f1a39b14007b4a62c1855d6e2f7348d17370c1dca2a4a17aa19cfeb7286b7427a95c9ca95d1190483075c33f893db9fc8526f59e53fee57f34673 to path /kaggle/input/bi-ml-competition-2023
Data source import complete.


# <center> Предсказание победителя в Dota 2
<center> <img src="https://meduza.io/impro/YnJZAHUW6WHz_JQm1uRPkTql_qAhbfxt3oFJLGH7CJg/fill/980/0/ce/1/aHR0cHM6Ly9tZWR1/emEuaW8vaW1hZ2Uv/YXR0YWNobWVudHMv/aW1hZ2VzLzAwNy8x/NTcvNjk1L29yaWdp/bmFsL0tMVThLbUti/ZG5pSzlibDA0Wmlw/WXcuanBn.webp" width="700" height="700">

[Почитать подбробнее](https://meduza.io/feature/2021/10/19/rossiyskaya-komanda-vyigrala-chempionat-mira-po-dota-2-i-poluchila-18-millionov-dollarov-postoyte-otkuda-takie-dengi-neuzheli-igrat-v-dotu-tak-slozhno)

#### [Оригинальная статья](https://arxiv.org/pdf/2106.01782.pdf)
    
### Начало

Посмотрим на готовые признаки и сделаем первую посылку.

1. [Описание данных](#Описание-данных)
2. [Описание признаков](#Описание-признаков)
3. [Наша первая модель](#Наша-первая-модель)
4. [Посылка](#Посылка)

### Первые шаги на пути в датасайенс

5. [Кросс-валидация](#Кросс-валидация)
6. [Что есть в json файлах?](#Что-есть-в-json-файлах?)
7. [Feature engineering](#Feature-engineering)

### Импорты

In [19]:
import os
import json
import pandas as pd
import datetime
import warnings
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split, ShuffleSplit, cross_val_score, GridSearchCV
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import roc_auc_score, accuracy_score
from lightgbm import LGBMClassifier
from catboost import CatBoostClassifier
%matplotlib inline

In [20]:
SEED = 10801
sns.set_style(style="whitegrid")
plt.rcParams["figure.figsize"] = 12, 8
warnings.filterwarnings("ignore")

## <left>Описание данных

Файлы:

- `sample_submission.csv`: пример файла-посылки
- `train_raw_data.jsonl`, `test_raw_data.jsonl`: "сырые" данные
- `train_data.csv`, `test_data.csv`: признаки, созданные авторами
- `train_targets.csv`: результаты тренировочных игр

## <left>Описание признаков
    
Набор простых признаков, описывающих игроков и команды в целом

In [21]:
PATH_TO_DATA = "/kaggle/input/copy-of-23-24-ml/"

df_train_features = pd.read_csv(os.path.join(PATH_TO_DATA,
                                             "train_data.csv"),
                                    index_col="match_id_hash")
df_train_targets = pd.read_csv(os.path.join(PATH_TO_DATA,
                                            "train_targets.csv"),
                                   index_col="match_id_hash")

In [22]:
df_train_features.shape

(31698, 245)

In [23]:
df_train_features.head()

game_time  game_mode  lobby_type  \
match_id_hash                                                        
b9c57c450ce74a2af79c9ce96fac144d        658          4           0   
6db558535151ea18ca70a6892197db41         21         23           0   
19c39fe2af2b547e48708ca005c6ae74        160         22           7   
c96d629dc0c39f0c616d1949938a6ba6       1016         22           0   
156c88bff4e9c4668b0f53df3d870f1b        582         22           7   

                                  objectives_len  chat_len  r1_hero_id  \
match_id_hash                                                            
b9c57c450ce74a2af79c9ce96fac144d               3        10          15   
6db558535151ea18ca70a6892197db41               0         0         101   
19c39fe2af2b547e48708ca005c6ae74               0         0          57   
c96d629dc0c39f0c616d1949938a6ba6               1         0         119   
156c88bff4e9c4668b0f53df3d870f1b               2         2          12   

                                  r1_kills  r1_deaths  r1_assists  r1_denies  \
match_id_hash                                                                  
b9c57c450ce74a2af79c9ce96fac144d         7          2           0          7   
6db558535151ea18ca70a6892197db41         0          0           0          0   
19c39fe2af2b547e48708ca005c6ae74         0          0           0          1   
c96d629dc0c39f0c616d1949938a6ba6         0          3           3          5   
156c88bff4e9c4668b0f53df3d870f1b         3          1           2          9   

                                  ...   d5_stuns  d5_creeps_stacked  \
match_id_hash                     ...                                 
b9c57c450ce74a2af79c9ce96fac144d  ...   0.000000                  0   
6db558535151ea18ca70a6892197db41  ...   0.000000                  0   
19c39fe2af2b547e48708ca005c6ae74  ...   0.000000                  0   
c96d629dc0c39f0c616d1949938a6ba6  ...   8.264696                  0   
156c88bff4e9c4668b0f53df3d870f1b  ...  15.762911                  3   

                                  d5_camps_stacked  d5_rune_pickups  \
match_id_hash                                                         
b9c57c450ce74a2af79c9ce96fac144d                 0                0   
6db558535151ea18ca70a6892197db41                 0                0   
19c39fe2af2b547e48708ca005c6ae74                 0                0   
c96d629dc0c39f0c616d1949938a6ba6                 0                3   
156c88bff4e9c4668b0f53df3d870f1b                 1                0   

                                  d5_firstblood_claimed  \
match_id_hash                                             
b9c57c450ce74a2af79c9ce96fac144d                      0   
6db558535151ea18ca70a6892197db41                      0   
19c39fe2af2b547e48708ca005c6ae74                      0   
c96d629dc0c39f0c616d1949938a6ba6                      0   
156c88bff4e9c4668b0f53df3d870f1b                      1   

                                  d5_teamfight_participation  \
match_id_hash                                                  
b9c57c450ce74a2af79c9ce96fac144d                        0.00   
6db558535151ea18ca70a6892197db41                        0.00   
19c39fe2af2b547e48708ca005c6ae74                        0.00   
c96d629dc0c39f0c616d1949938a6ba6                        0.25   
156c88bff4e9c4668b0f53df3d870f1b                        0.50   

                                  d5_towers_killed  d5_roshans_killed  \
match_id_hash                                                           
b9c57c450ce74a2af79c9ce96fac144d                 0                  0   
6db558535151ea18ca70a6892197db41                 0                  0   
19c39fe2af2b547e48708ca005c6ae74                 0                  0   
c96d629dc0c39f0c616d1949938a6ba6                 0                  0   
156c88bff4e9c4668b0f53df3d870f1b                 0                  0   

                                  d5_obs_placed  d5_sen_placed  
match_id_hash              

Имеем ~32 тысячи наблюдений, каждое из которых характеризуется уникальным `match_id_hash` (захэшированное id матча), и 245 признаков. `game_time` показывает момент времени, в который получены эти данные. То есть по сути это не длительность самого матча, а например, его середина, таким образом, в итоге мы сможем получить модель, которая будет предсказывать вероятность победы каждой из команд в течение матча (хорошо подходит для букмекеров).

Нас интересует поле `radiant_win` (так называется одна из команд, вторая - dire). Остальные колоки здесь по сути получены из "будущего" и есть только для тренировочных данных, поэтому на них можно просто посмотреть).

In [36]:
df_train_targets.head()

game_time  radiant_win  duration  \
match_id_hash                                                        
b9c57c450ce74a2af79c9ce96fac144d        658         True      1154   
6db558535151ea18ca70a6892197db41         21         True      1503   
19c39fe2af2b547e48708ca005c6ae74        160        False      2063   
c96d629dc0c39f0c616d1949938a6ba6       1016         True      2147   
156c88bff4e9c4668b0f53df3d870f1b        582        False      1927   

                                  time_remaining next_roshan_team  
match_id_hash                                                      
b9c57c450ce74a2af79c9ce96fac144d             496              NaN  
6db558535151ea18ca70a6892197db41            1482          Radiant  
19c39fe2af2b547e48708ca005c6ae74            1903              NaN  
c96d629dc0c39f0c616d1949938a6ba6            1131          Radiant  
156c88bff4e9c4668b0f53df3d870f1b            1345             Dire

## <left> Зададим train, target и 

In [37]:
X = df_train_features.values
y = df_train_targets["radiant_win"].values.astype("int8")

In [38]:
X_train, X_valid, y_train, y_valid = train_test_split(X, y,
                                                      test_size=0.3,
                                                      random_state=SEED)

In [39]:
try:
    import ujson as json
except ModuleNotFoundError:
    import json
    print ("Подумайте об установке ujson, чтобы работать с JSON объектами быстрее")

try:
    from tqdm.notebook import tqdm
except ModuleNotFoundError:
    tqdm_notebook = lambda x: x
    print ("Подумайте об установке tqdm, чтобы следить за прогрессом")


def read_matches(matches_file, total_matches=31698, n_matches_to_read=None):
    """
    Аргуент
    -------
    matches_file: JSON файл с сырыми данными

    Результат
    ---------
    Возвращает записи о каждом матче
    """

    if n_matches_to_read is None:
        n_matches_to_read = total_matches

    c = 0
    with open(matches_file) as fin:
        for line in tqdm(fin, total=total_matches):
            if c >= n_matches_to_read:
                break
            else:
                c += 1
                yield json.loads(line)

In [40]:
def add_new_features(df_features, matches_file):
    """
    Аргуенты
    -------
    df_features: таблица с данными
    matches_file: JSON файл с сырыми данными

    Результат
    ---------
    Добавляет новые признаки в таблицу
    """

    for match in read_matches(matches_file):
        match_id_hash = match['match_id_hash']

        # Посчитаем количество разрушенных вышек обеими командами
        radiant_tower_kills = 0
        dire_tower_kills = 0
        radiant_roshan_kills = 0
        dire_roshan_kills = 0
        for objective in match["objectives"]:
            if objective["type"] == "CHAT_MESSAGE_TOWER_KILL":
                if objective["team"] == 2:
                    radiant_tower_kills += 1
                if objective["team"] == 3:
                    dire_tower_kills += 1
            if objective["type"] == "CHAT_MESSAGE_ROSHAN_KILL":
                if objective["team"] == 2:
                    radiant_roshan_kills += 1
                if objective["team"] == 3:
                    dire_roshan_kills += 1
                

        df_features.loc[match_id_hash, "radiant_tower_kills"] = radiant_tower_kills
        df_features.loc[match_id_hash, "dire_tower_kills"] = dire_tower_kills
        df_features.loc[match_id_hash, "diff_tower_kills"] = radiant_tower_kills - dire_tower_kills
        
        df_features.loc[match_id_hash, "radiant_roshan_kills"] = radiant_roshan_kills
        df_features.loc[match_id_hash, "dire_roshan_kills"] = dire_roshan_kills
        df_features.loc[match_id_hash, "diff_roshan_kills"] = radiant_roshan_kills - dire_roshan_kills
        

        # ... (/¯◡ ‿ ◡)/¯☆*:・ﾟ добавляем новые признаки ...

In [41]:
try:
    import ujson as json
except ModuleNotFoundError:
    import json
    print ("Подумайте об установке ujson, чтобы работать с JSON объектами быстрее")

try:
    from tqdm.notebook import tqdm
except ModuleNotFoundError:
    tqdm_notebook = lambda x: x
    print ("Подумайте об установке tqdm, чтобы следить за прогрессом")


def read_matches(matches_file, total_matches=31698, n_matches_to_read=None):
    """
    Аргуент
    -------
    matches_file: JSON файл с сырыми данными

    Результат
    ---------
    Возвращает записи о каждом матче
    """

    if n_matches_to_read is None:
        n_matches_to_read = total_matches

    c = 0
    with open(matches_file) as fin:
        for line in tqdm(fin, total=total_matches):
            if c >= n_matches_to_read:
                break
            else:
                c += 1
                yield json.loads(line)

In [42]:
def read_matches(matches_file, total_matches=31698, n_matches_to_read=None):
    """
    Аргуент
    -------
    matches_file: JSON файл с сырыми данными

    Результат
    ---------
    Возвращает записи о каждом матче
    """

    if n_matches_to_read is None:
        n_matches_to_read = total_matches

    c = 0
    with open(matches_file) as fin:
        for line in tqdm(fin, total=total_matches):
            if c >= n_matches_to_read:
                break
            else:
                c += 1
                yield json.loads(line)

In [43]:
# Скопируем таблицу с признаками
df_train_features_extended = df_train_features.copy()

# Добавим новые
add_new_features(df_train_features_extended,
                 os.path.join(PATH_TO_DATA,
                              "train_raw_data.jsonl"))

  0%|          | 0/31698 [00:00<?, ?it/s]

In [48]:
df_train_features_extended

game_time  game_mode  lobby_type  \
match_id_hash                                                        
b9c57c450ce74a2af79c9ce96fac144d        658          4           0   
6db558535151ea18ca70a6892197db41         21         23           0   
19c39fe2af2b547e48708ca005c6ae74        160         22           7   
c96d629dc0c39f0c616d1949938a6ba6       1016         22           0   
156c88bff4e9c4668b0f53df3d870f1b        582         22           7   
...                                     ...        ...         ...   
b794aa55646841a03c7783226e6f6bc8       2666         22           0   
308faee28efee2e66b39f9f2ba6ea9cf       2525         22           0   
6066cc7417b43c749d551e123d00f0c8       1002          4           0   
bc7a87ed5f9c2bca55f9f7a93da0b0c5        377         22           7   
e2ca68ac1a6847f4a37f6c9c8ee8695b        643         22           7   

                                  objectives_len  chat_len  r1_hero_id  \
match_id_hash                                                            
b9c57c450ce74a2af79c9ce96fac144d               3        10          15   
6db558535151ea18ca70a6892197db41               0         0         101   
19c39fe2af2b547e48708ca005c6ae74               0         0          57   
c96d629dc0c39f0c616d1949938a6ba6               1         0         119   
156c88bff4e9c4668b0f53df3d870f1b               2         2          12   
...                                          ...       ...         ...   
b794aa55646841a03c7783226e6f6bc8              23         8          99   
308faee28efee2e66b39f9f2ba6ea9cf              15         5          22   
6066cc7417b43c749d551e123d00f0c8               4         0           1   
bc7a87ed5f9c2bca55f9f7a93da0b0c5               1         0          14   
e2ca68ac1a6847f4a37f6c9c8ee8695b               1        23          63   

                                  r1_kills  r1_deaths  r1_assists  r1_denies  \
match_id_hash                                                                  
b9c57c450ce74a2af79c9ce96fac144d         7          2           0          7   
6db558535151ea18ca70a6892197db41         0          0           0          0   
19c39fe2af2b547e48708ca005c6ae74         0          0           0          1   
c96d629dc0c39f0c616d1949938a6ba6         0          3           3          5   
156c88bff4e9c4668b0f53df3d870f1b         3          1           2          9   
...                                    ...        ...         ...        ...   
b794aa55646841a03c7783226e6f6bc8         9         11           9          3   
308faee28efee2e66b39f9f2ba6ea9cf         3          9          14          3   
6066cc7417b43c749d551e123d00f0c8         1          2           1          0   
bc7a87ed5f9c2bca55f9f7a93da0b0c5         0          2           1          1   
e2ca68ac1a6847f4a37f6c9c8ee8695b         1          4           0          2   

                                  ...  d5_towers_killed  d5_roshans_killed  \
match_id_hash                     ...                                        
b9c57c450ce74a2af79c9ce96fac144d  ...                 0                  0   
6db558535151ea18ca70a6892197db41  ...                 0                  0   
19c39fe2af2b547e48708ca005c6ae74  ...                 0                  0   
c96d629dc0c39f0c616d1949938a6ba6  ...                 0                  0   
156c88bff4e9c4668b0f53df3d870f1b  ...                 0                  0   
...                               ...               ...                ...   
b794aa55646841a03c7783226e6f6bc8  ...                 2                  2   
308faee28efee2e66b39f9f2ba6ea9cf  ...                 0                  0   
6066cc7417b43c749d551e123d00f0c8  ...                 0                  0   
bc7a87ed5f9c2bca55f9f7a93da0b0c5  ...                 0                  0   
e2ca68ac1a6847f4a37f6c9c8ee8695b  ...                 0                  0   

                                  d5_obs_placed  d5_sen_placed  \
match_id_hash                         

In [49]:
# Считаем сумму золота для команды Radiant
df_train_features_extended['r_total_gold'] = df_train_features_extended[['r1_gold', 'r2_gold', 'r3_gold', 'r4_gold', 'r5_gold']].sum(axis=1)

# Считаем сумму золота для команды Dire
df_train_features_extended['d_total_gold'] = df_train_features_extended[['d1_gold', 'd2_gold', 'd3_gold', 'd4_gold', 'd5_gold']].sum(axis=1)

# Вычисляем разницу между суммами золота команд
df_train_features_extended['gold_diff'] = df_train_features_extended['r_total_gold'] - df_train_features_extended['d_total_gold']

# Вычисляем разницу между суммами золота команд, деленную на время игры
df_train_features_extended['gold_diff_per_second'] = df_train_features_extended['gold_diff'] / df_train_features_extended['game_time']


In [50]:
df_train_features_extended

game_time  game_mode  lobby_type  \
match_id_hash                                                        
b9c57c450ce74a2af79c9ce96fac144d        658          4           0   
6db558535151ea18ca70a6892197db41         21         23           0   
19c39fe2af2b547e48708ca005c6ae74        160         22           7   
c96d629dc0c39f0c616d1949938a6ba6       1016         22           0   
156c88bff4e9c4668b0f53df3d870f1b        582         22           7   
...                                     ...        ...         ...   
b794aa55646841a03c7783226e6f6bc8       2666         22           0   
308faee28efee2e66b39f9f2ba6ea9cf       2525         22           0   
6066cc7417b43c749d551e123d00f0c8       1002          4           0   
bc7a87ed5f9c2bca55f9f7a93da0b0c5        377         22           7   
e2ca68ac1a6847f4a37f6c9c8ee8695b        643         22           7   

                                  objectives_len  chat_len  r1_hero_id  \
match_id_hash                                                            
b9c57c450ce74a2af79c9ce96fac144d               3        10          15   
6db558535151ea18ca70a6892197db41               0         0         101   
19c39fe2af2b547e48708ca005c6ae74               0         0          57   
c96d629dc0c39f0c616d1949938a6ba6               1         0         119   
156c88bff4e9c4668b0f53df3d870f1b               2         2          12   
...                                          ...       ...         ...   
b794aa55646841a03c7783226e6f6bc8              23         8          99   
308faee28efee2e66b39f9f2ba6ea9cf              15         5          22   
6066cc7417b43c749d551e123d00f0c8               4         0           1   
bc7a87ed5f9c2bca55f9f7a93da0b0c5               1         0          14   
e2ca68ac1a6847f4a37f6c9c8ee8695b               1        23          63   

                                  r1_kills  r1_deaths  r1_assists  r1_denies  \
match_id_hash                                                                  
b9c57c450ce74a2af79c9ce96fac144d         7          2           0          7   
6db558535151ea18ca70a6892197db41         0          0           0          0   
19c39fe2af2b547e48708ca005c6ae74         0          0           0          1   
c96d629dc0c39f0c616d1949938a6ba6         0          3           3          5   
156c88bff4e9c4668b0f53df3d870f1b         3          1           2          9   
...                                    ...        ...         ...        ...   
b794aa55646841a03c7783226e6f6bc8         9         11           9          3   
308faee28efee2e66b39f9f2ba6ea9cf         3          9          14          3   
6066cc7417b43c749d551e123d00f0c8         1          2           1          0   
bc7a87ed5f9c2bca55f9f7a93da0b0c5         0          2           1          1   
e2ca68ac1a6847f4a37f6c9c8ee8695b         1          4           0          2   

                                  ...  radiant_tower_kills  dire_tower_kills  \
match_id_hash                     ...                                          
b9c57c450ce74a2af79c9ce96fac144d  ...                  2.0               0.0   
6db558535151ea18ca70a6892197db41  ...                  0.0               0.0   
19c39fe2af2b547e48708ca005c6ae74  ...                  0.0               0.0   
c96d629dc0c39f0c616d1949938a6ba6  ...                  0.0               0.0   
156c88bff4e9c4668b0f53df3d870f1b  ...                  0.0               1.0   
...                               ...                  ...               ...   
b794aa55646841a03c7783226e6f6bc8  ...                  4.0               8.0   
308faee28efee2e66b39f9f2ba6ea9cf  ...                  6.0               4.0   
6066cc7417b43c749d551e123d00f0c8  ...                  2.0               1.0   
bc7a87ed5f9c2bca55f9f7a93da0b0c5  ...                  0.0               0.0   
e2ca68ac1a6847f4a37f6c9c8ee8695b  ...                  0.0               0.0   

                                  diff_tower_kills  radiant_roshan_kills  \
ma

In [51]:
from sklearn.model_selection import StratifiedKFold


In [52]:
cv_5 = StratifiedKFold(n_splits=5)

In [53]:
# Получаем индексы строк, где 'game_time' больше 600
indices = df_train_features_extended['game_time'] > 600# Фильтруем df_train_features_extended и y с использованием этих индексов
df_train_features_after_10min = df_train_features_extended.loc[indices]
y_after_10min = y[indices]
rf_model = RandomForestClassifier(n_estimators=300, max_depth=7, n_jobs=-1, random_state=SEED)

In [54]:
# Обучаем модель на данных, исключающих матчи с замерами до 10 минут
rf_model.fit(df_train_features_after_10min.values, y_after_10min)

RandomForestClassifier(max_depth=7, n_estimators=300, n_jobs=-1,
                       random_state=10801)

In [55]:
%%time
cv_scores_base = cross_val_score(rf_model, X, y, cv=cv_5, scoring="roc_auc", n_jobs=-1)
cv_scores_extended_10 = cross_val_score(rf_model, df_train_features_after_10min.values, y_after_10min,
                                     cv=cv_5, scoring="roc_auc", n_jobs=-1)

CPU times: total: 219 ms
Wall time: 30.7 s


In [56]:
print(f"ROC-AUC на кросс-валидации для матчей >600sec: {cv_scores_extended_10.mean()}")

ROC-AUC на кросс-валидации для матчей >600sec: 0.8475455918452368


### Посылаем

In [19]:
X_test = df_test_features_extended.values
y_test_pred = rf_model.predict_proba(X_test)[:, 1]

df_submission = pd.DataFrame({"radiant_win_prob": y_test_pred},
                                index=df_test_features.index)

NameError: name 'df_test_features_extended' is not defined

In [ ]:
submission_filename = "submission_{}.csv".format(
    datetime.datetime.now().strftime("%Y-%m-%d_%H-%M-%S"))
df_submission.to_csv(submission_filename)
print("Файл посылки сохранен, как: {}".format(submission_filename))

In [57]:

# Создаем экземпляр класса LGBMClassifier
lgb_model = LGBMClassifier(n_estimators=300, max_depth=7, num_leaves=2**7, random_state=SEED)

# Обучаем модель на данных, исключающих матчи с замерами до 10 минут
lgb_model.fit(df_train_features_after_10min.values, y_after_10min)

# Вычисляем ROC-AUC на кросс-валидации
cv_scores_lgb = cross_val_score(lgb_model, df_train_features_after_10min.values, y_after_10min,
                                cv=cv_5, scoring="roc_auc", n_jobs=-1)

print(f"ROC-AUC на кросс-валидации для матчей >600sec (LightGBM): {cv_scores_lgb.mean()}")


[LightGBM] [Info] Number of positive: 11764, number of negative: 10857
[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.013060 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 27773
[LightGBM] [Info] Number of data points in the train set: 22621, number of used features: 255
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.520048 -> initscore=0.080234
[LightGBM] [Info] Start training from score 0.080234
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gai

In [58]:
# Вычисляем ROC-AUC на кросс-валидации
cv_scores_lgb = cross_val_score(lgb_model, df_train_features_extended.values, y,
                                cv=cv_5, scoring="roc_auc", n_jobs=-1)

print(f"ROC-AUC на кросс-валидации для матчей >600sec (LightGBM): {cv_scores_lgb.mean()}")

ROC-AUC на кросс-валидации для матчей >600sec (LightGBM): 0.8102425986916625


In [59]:

# Создаем экземпляр класса CatBoostClassifier
cb_model = CatBoostClassifier(iterations=300, depth=7, verbose=False, random_state=SEED)

# Обучаем модель на данных, исключающих матчи с замерами до 10 минут
cb_model.fit(df_train_features_after_10min.values, y_after_10min)

# Вычисляем ROC-AUC на кросс-валидации
cv_scores_cb = cross_val_score(cb_model, df_train_features_after_10min.values, y_after_10min,
                               cv=cv_5, scoring="roc_auc", n_jobs=-1)

print(f"ROC-AUC на кросс-валидации для матчей >600sec (CatBoost): {cv_scores_cb.mean()}")


ROC-AUC на кросс-валидации для матчей >600sec (CatBoost): 0.8500110331185574


In [60]:
# Вычисляем ROC-AUC на кросс-валидации
cv_scores_cb = cross_val_score(cb_model, df_train_features_extended.values, y,
                               cv=cv_5, scoring="roc_auc", n_jobs=-1)

print(f"ROC-AUC на кросс-валидации для матчей >600sec (CatBoost): {cv_scores_cb.mean()}")

ROC-AUC на кросс-валидации для матчей >600sec (CatBoost): 0.810261314341693


In [ ]:


# Определяем модель CatBoost
cb_model = CatBoostClassifier(verbose=False)

# Задаем сетку параметров, которые хотим проверить
param_grid = {
    'depth': [5, 7, 9],
    'iterations': [50, 100],
    'learning_rate': [0.01, 0.05, 0.1],
    'l2_leaf_reg': [1, 3, 5],
    'border_count': [32, 64, 128]
}


# Создаем объект GridSearchCV
grid_search = GridSearchCV(estimator=cb_model, param_grid=param_grid, cv=3, scoring='roc_auc')

# Выполняем решетчатый поиск на обучающих данных
grid_search.fit(X_train, y_train)

# Выводим наилучшие параметры
print("Наилучшие параметры: ", grid_search.best_params_)


In [78]:
from catboost import CatBoostClassifier

# Замените эти значения на те, которые были найдены в результате решетчатого поиска
best_params = {'depth': 6, 'iterations': 100, 'learning_rate': 0.1}

# Создаем экземпляр класса CatBoostClassifier с лучшими параметрами
cb_model = CatBoostClassifier(**best_params, random_state=SEED)

# Обучаем модель на данных
cb_model.fit(X_train, y_train)


0:	learn: 0.6795394	total: 14.5ms	remaining: 1.44s
1:	learn: 0.6670205	total: 26.8ms	remaining: 1.31s
2:	learn: 0.6559735	total: 39.8ms	remaining: 1.29s
3:	learn: 0.6476567	total: 54.2ms	remaining: 1.3s
4:	learn: 0.6392241	total: 67.4ms	remaining: 1.28s
5:	learn: 0.6327256	total: 79.6ms	remaining: 1.25s
6:	learn: 0.6270337	total: 92.5ms	remaining: 1.23s
7:	learn: 0.6219201	total: 105ms	remaining: 1.2s
8:	learn: 0.6161027	total: 118ms	remaining: 1.2s
9:	learn: 0.6110846	total: 131ms	remaining: 1.18s
10:	learn: 0.6070361	total: 144ms	remaining: 1.16s
11:	learn: 0.6034091	total: 156ms	remaining: 1.15s
12:	learn: 0.5994462	total: 169ms	remaining: 1.13s
13:	learn: 0.5966123	total: 181ms	remaining: 1.11s
14:	learn: 0.5930368	total: 194ms	remaining: 1.1s
15:	learn: 0.5901135	total: 207ms	remaining: 1.08s
16:	learn: 0.5875077	total: 220ms	remaining: 1.07s
17:	learn: 0.5847571	total: 235ms	remaining: 1.07s
18:	learn: 0.5818919	total: 249ms	remaining: 1.06s
19:	learn: 0.5801928	total: 262ms	rema

In [79]:
# Вычисляем ROC-AUC на кросс-валидации
cv_scores_cb = cross_val_score(cb_model, df_train_features_extended.values, y,
                               cv=cv_5, scoring="roc_auc", n_jobs=-1)

print(f"ROC-AUC на кросс-валидации для матчей >600sec (CatBoost): {cv_scores_cb.mean()}")

ROC-AUC на кросс-валидации для матчей >600sec (CatBoost): 0.8115341088460631


In [61]:
df_test_features = pd.read_csv(os.path.join(PATH_TO_DATA, "test_data.csv"),
                                   index_col="match_id_hash")

In [62]:
# Скопируем таблицу с признаками
df_test_features_extended = df_test_features.copy()

# Добавим новые
add_new_features(df_test_features_extended,
                 os.path.join(PATH_TO_DATA,
                              "test_raw_data.jsonl"))

  0%|          | 0/31698 [00:00<?, ?it/s]

In [70]:
df_test_features_extended

game_time  game_mode  lobby_type  \
match_id_hash                                                        
a400b8f29dece5f4d266f49f1ae2e98a        155         22           7   
46a0ddce8f7ed2a8d9bd5edcbb925682        576         22           7   
b1b35ff97723d9b7ade1c9c3cf48f770        453         22           7   
ab3cc6ccac661a1385e73a2e9f21313a        721          4           0   
54aaab1cb8cc5df3c253641618673266        752         22           7   
...                                     ...        ...         ...   
03f683fed8cf6f03b735e2ee2a942698        509          4           0   
06e2b42495855647830ccc6d466f913c       2561         22           7   
defd0caeed6ea83d7d5fbdec013fe7d1       1783         22           0   
47ad6454ede66c1c78fdaa9391dfc556       2405         22           7   
9928dfde50efcbdb2055da23dcdbc101       1775         22           0   

                                  objectives_len  chat_len  r1_hero_id  \
match_id_hash                                                            
a400b8f29dece5f4d266f49f1ae2e98a               1        11          11   
46a0ddce8f7ed2a8d9bd5edcbb925682               1         4          14   
b1b35ff97723d9b7ade1c9c3cf48f770               1         3          42   
ab3cc6ccac661a1385e73a2e9f21313a               2         1          30   
54aaab1cb8cc5df3c253641618673266               1         0           8   
...                                          ...       ...         ...   
03f683fed8cf6f03b735e2ee2a942698               1         0          16   
06e2b42495855647830ccc6d466f913c              17         7          76   
defd0caeed6ea83d7d5fbdec013fe7d1               8        23           3   
47ad6454ede66c1c78fdaa9391dfc556              12         4          22   
9928dfde50efcbdb2055da23dcdbc101              10        13          32   

                                  r1_kills  r1_deaths  r1_assists  r1_denies  \
match_id_hash                                                                  
a400b8f29dece5f4d266f49f1ae2e98a         0          0           0          0   
46a0ddce8f7ed2a8d9bd5edcbb925682         1          0           3          1   
b1b35ff97723d9b7ade1c9c3cf48f770         0          1           1          0   
ab3cc6ccac661a1385e73a2e9f21313a         2          2           1          3   
54aaab1cb8cc5df3c253641618673266         2          0           2          8   
...                                    ...        ...         ...        ...   
03f683fed8cf6f03b735e2ee2a942698         2          0           0          1   
06e2b42495855647830ccc6d466f913c         2          9           1          6   
defd0caeed6ea83d7d5fbdec013fe7d1         1          9           9          0   
47ad6454ede66c1c78fdaa9391dfc556         3          8          14          7   
9928dfde50efcbdb2055da23dcdbc101         3          4          15          0   

                                  ...  d5_towers_killed  d5_roshans_killed  \
match_id_hash                     ...                                        
a400b8f29dece5f4d266f49f1ae2e98a  ...                 0                  0   
46a0ddce8f7ed2a8d9bd5edcbb925682  ...                 0                  0   
b1b35ff97723d9b7ade1c9c3cf48f770  ...                 0                  0   
ab3cc6ccac661a1385e73a2e9f21313a  ...                 0                  0   
54aaab1cb8cc5df3c253641618673266  ...                 0                  0   
...                               ...               ...                ...   
03f683fed8cf6f03b735e2ee2a942698  ...                 0                  0   
06e2b42495855647830ccc6d466f913c  ...                 0                  0   
defd0caeed6ea83d7d5fbdec013fe7d1  ...                 0                  0   
47ad6454ede66c1c78fdaa9391dfc556  ...                 0                  0   
9928dfde50efcbdb2055da23dcdbc101  ...                 1                  0   

                                  d5_obs_placed  d5_sen_placed  \
match_id_hash                         

In [71]:
# Считаем сумму золота для команды Radiant
df_test_features_extended['r_total_gold'] = df_test_features_extended[['r1_gold', 'r2_gold', 'r3_gold', 'r4_gold', 'r5_gold']].sum(axis=1)

# Считаем сумму золота для команды Dire
df_test_features_extended['d_total_gold'] = df_test_features_extended[['d1_gold', 'd2_gold', 'd3_gold', 'd4_gold', 'd5_gold']].sum(axis=1)

# Вычисляем разницу между суммами золота команд
df_test_features_extended['gold_diff'] = df_test_features_extended['r_total_gold'] - df_test_features_extended['d_total_gold']

# Вычисляем разницу между суммами золота команд, деленную на время игры
df_test_features_extended['gold_diff_per_second'] = df_test_features_extended['gold_diff'] / df_test_features_extended['game_time']

# Переводим обновленный DataFrame в массив NumPy
X_test = df_test_features_extended.values


In [72]:
df_test_features_extended

game_time  game_mode  lobby_type  \
match_id_hash                                                        
a400b8f29dece5f4d266f49f1ae2e98a        155         22           7   
46a0ddce8f7ed2a8d9bd5edcbb925682        576         22           7   
b1b35ff97723d9b7ade1c9c3cf48f770        453         22           7   
ab3cc6ccac661a1385e73a2e9f21313a        721          4           0   
54aaab1cb8cc5df3c253641618673266        752         22           7   
...                                     ...        ...         ...   
03f683fed8cf6f03b735e2ee2a942698        509          4           0   
06e2b42495855647830ccc6d466f913c       2561         22           7   
defd0caeed6ea83d7d5fbdec013fe7d1       1783         22           0   
47ad6454ede66c1c78fdaa9391dfc556       2405         22           7   
9928dfde50efcbdb2055da23dcdbc101       1775         22           0   

                                  objectives_len  chat_len  r1_hero_id  \
match_id_hash                                                            
a400b8f29dece5f4d266f49f1ae2e98a               1        11          11   
46a0ddce8f7ed2a8d9bd5edcbb925682               1         4          14   
b1b35ff97723d9b7ade1c9c3cf48f770               1         3          42   
ab3cc6ccac661a1385e73a2e9f21313a               2         1          30   
54aaab1cb8cc5df3c253641618673266               1         0           8   
...                                          ...       ...         ...   
03f683fed8cf6f03b735e2ee2a942698               1         0          16   
06e2b42495855647830ccc6d466f913c              17         7          76   
defd0caeed6ea83d7d5fbdec013fe7d1               8        23           3   
47ad6454ede66c1c78fdaa9391dfc556              12         4          22   
9928dfde50efcbdb2055da23dcdbc101              10        13          32   

                                  r1_kills  r1_deaths  r1_assists  r1_denies  \
match_id_hash                                                                  
a400b8f29dece5f4d266f49f1ae2e98a         0          0           0          0   
46a0ddce8f7ed2a8d9bd5edcbb925682         1          0           3          1   
b1b35ff97723d9b7ade1c9c3cf48f770         0          1           1          0   
ab3cc6ccac661a1385e73a2e9f21313a         2          2           1          3   
54aaab1cb8cc5df3c253641618673266         2          0           2          8   
...                                    ...        ...         ...        ...   
03f683fed8cf6f03b735e2ee2a942698         2          0           0          1   
06e2b42495855647830ccc6d466f913c         2          9           1          6   
defd0caeed6ea83d7d5fbdec013fe7d1         1          9           9          0   
47ad6454ede66c1c78fdaa9391dfc556         3          8          14          7   
9928dfde50efcbdb2055da23dcdbc101         3          4          15          0   

                                  ...  radiant_tower_kills  dire_tower_kills  \
match_id_hash                     ...                                          
a400b8f29dece5f4d266f49f1ae2e98a  ...                  0.0               0.0   
46a0ddce8f7ed2a8d9bd5edcbb925682  ...                  0.0               0.0   
b1b35ff97723d9b7ade1c9c3cf48f770  ...                  0.0               0.0   
ab3cc6ccac661a1385e73a2e9f21313a  ...                  1.0               0.0   
54aaab1cb8cc5df3c253641618673266  ...                  0.0               0.0   
...                               ...                  ...               ...   
03f683fed8cf6f03b735e2ee2a942698  ...                  0.0               0.0   
06e2b42495855647830ccc6d466f913c  ...                  3.0               7.0   
defd0caeed6ea83d7d5fbdec013fe7d1  ...                  3.0               4.0   
47ad6454ede66c1c78fdaa9391dfc556  ...                  6.0               3.0   
9928dfde50efcbdb2055da23dcdbc101  ...                  3.0               5.0   

                                  diff_tower_kills  radiant_roshan_kills  \
ma

In [80]:
X_test = df_test_features_extended.values

In [81]:
y_test_pred = cb_model.predict_proba(X_test)[:, 1]

In [82]:
df_submission = pd.DataFrame({"radiant_win_prob": y_test_pred},
                                 index=df_test_features.index)

In [83]:
submission_filename = "submission_{}.csv".format(
    datetime.datetime.now().strftime("%Y-%m-%d_%H-%M-%S"))
df_submission.to_csv(submission_filename)
print("Файл посылки сохранен, как: {}".format(submission_filename))

Файл посылки сохранен, как: submission_2024-04-05_19-36-16.csv
